File: SDKs/Aspose_Cloud_SDK_For_Android/asposecloudsdk/src/main/java/com/aspose/cloud/sdk/words/api/Converter.java
package com.aspose.cloud.sdk.words.api;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;

import android.net.Uri;

import com.aspose.cloud.sdk.common.AsposeApp;
import com.aspose.cloud.sdk.common.Utils;
import com.aspose.cloud.sdk.words.model.ConvertWebPageToWordDocumentResponse;
import com.aspose.cloud.sdk.words.model.ConvertWordDocumentToAnyFormatWithAdditionalSettingsResponse;
import com.aspose.cloud.sdk.words.model.SaveResult;
import com.aspose.cloud.sdk.words.model.ValidFormatsEnum;
import com.google.gson.Gson;

/**
 * Converter --- Using this class you can convert a Word document to images, multipage tiff, HTML, PDF and other file formats
 * on the Aspose cloud storage and without using cloud Storage.
 * @author   M. Sohail Ismail
 */

public class Converter {
	private static final String WORD_URI = AsposeApp.BASE_PRODUCT_URI + "/words/";
	
	/**
	 * Convert a Word document to images, multipage tiff, HTML, PDF and other file formats.
	 * @param fileName Name of the MS Word document on cloud
	 * @param designatedFormat A format to which word docuemnt will be converted
	 * @throws java.security.InvalidKeyException If initialization fails because the provided key is null.
	 * @throws java.security.NoSuchAlgorithmException If the specified algorithm (HmacSHA1) is not available by any provider.
	 * @throws java.io.IOException If there is an IO error
	 * @return A path to converted word document
	*/ 
	public static String convertWordDocumentToFormat(String fileName, ValidFormatsEnum designatedFormat) throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		
		String localFilePath = null;
		
		if(fileName == null || fileName.length() <= 3) {
			throw new IllegalArgumentException("File name cannot be null or empty");
		}
		
		if(designatedFormat == null) {
			throw new IllegalArgumentException("Designated format cannot be null");
		}
		
		//build URL
		String strURL = WORD_URI + Uri.encode(fileName) + "?format=" + designatedFormat;
		//sign URL
		String signedURL = Utils.sign(strURL);
		InputStream responseStream = Utils.processCommand(signedURL, "GET");
		
		//Replace fileName extension with designated format 
		String[] fileNameAndItsExtensionArray = fileName.split("\\.");
		fileName = fileNameAndItsExtensionArray[0] + "." + designatedFormat;
		
		//Save file on Disk
		localFilePath = Utils.saveStreamToFile(responseStream, fileName);
		return localFilePath;
	}
	
	/**
	 * Convert Word document that is locally stored on device to images, multipage tiff, HTML, PDF and other file formats.
	 * @param localFilePath A path to Word document on disk
	 * @param designatedFormat A format to which word docuemnt will be converted
	 * @throws java.security.InvalidKeyException If initialization fails because the provided key is null.
	 * @throws java.security.NoSuchAlgorithmException If the specified algorithm (HmacSHA1) is not available by any provider.
	 * @throws java.io.IOException If there is an IO error
	 * @return A path to converted word document
	*/ 
	public static String convertLocallyStoredWordDocumentToFormat(String localFilePath, ValidFormatsEnum designatedFormat) throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		String convertedFilePath = null;
		
		if(localFilePath == null || localFilePath.length() == 0) {
			throw new IllegalArgumentException("Local file path cannot be null or empty");
		}
		
		if(designatedFormat == null) {
			throw new IllegalArgumentException("Designated format cannot be null");
		}
		
		//Build the request URI
		String strURL = WORD_URI + "convert?format=" + designatedFormat;
		//Sign the request URI
		String signedURL = Utils.sign(strURL);	
		//Convert the local file to InputStream
		InputStream fileStream = new FileInputStream(localFilePath);
		//Process the request on server
		InputStream responseStream = Utils.processCommand(signedURL, "PUT", fileStream);
		//Get fileName from localFilePath
		String fileName;
		int index = localFilePath.lastIndexOf("/");
		if(index != -1) {
			fileName = localFilePath.substring(index+1);
		} else {
			fileName = localFilePath;
		}
		//Replace fileName extension with designated format 
		String[] fileNameAndItsExtensionArray = fileName.split("\\.");
		fileName = fileNameAndItsExtensionArray[0] + "." + designatedFormat;
				
		//Save file on Disk
		convertedFilePath = Utils.saveStreamToFile(responseStream, fileName);
		
		return convertedFilePath;
	}
	
	/**
	 * Convert a Word document to other formats with additional settings
	 * @param fileName Name of the MS Word document on cloud
	 * @param saveOptionsRequest Particular file format save options
	 * @param dataInterchangeFormat Can be either XML or JSON
	 * @throws java.security.InvalidKeyException If initialization fails because the provided key is null.
	 * @throws java.security.NoSuchAlgorithmException If the specified algorithm (HmacSHA1) is not available by any provider.
	 * @throws java.io.IOException If there is an IO error
	 * @return A path to converted word document
	*/
	public static SaveResult convertWordDocumentToFormatWithAdditionalSettings(String fileName, String saveOptionsRequest, String dataInterchangeFormat) throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		SaveResult saveResult = null;
		
		if(fileName == null || fileName.length() <= 3) {
			throw new IllegalArgumentException("File name cannot be null or empty");
		}
		
		if(saveOptionsRequest == null || saveOptionsRequest.length() == 0) {
			throw new IllegalArgumentException("Save options request cannot be null or empty");
		}
		
		if(dataInterchangeFormat == null || dataInterchangeFormat.length() == 0) {
			throw new IllegalArgumentException("Data interchange format cannot be null or empty. Either it should be XML or JSON");
		}
		
		//build URL
		String strURL = WORD_URI + Uri.encode(fileName) + "/saveAs";
		//sign URL
		String signedURL = Utils.sign(strURL);
		InputStream responseStream = Utils.processCommand(signedURL, "POST", saveOptionsRequest, dataInterchangeFormat);
		String responseJSONString = Utils.streamToString(responseStream);
		
		//Parsing JSON
      	Gson gson = new Gson();
      	ConvertWordDocumentToAnyFormatWithAdditionalSettingsResponse convertWordDocResponse = gson.fromJson(responseJSONString, ConvertWordDocumentToAnyFormatWithAdditionalSettingsResponse.class);
		if(convertWordDocResponse.getCode().equals("200") && convertWordDocResponse.getStatus().equals("OK")) {
			saveResult = convertWordDocResponse.saveResult;
		}
		
		return saveResult;
	}
	
	/**
	 * Convert a live web page to Word document
	 * @param loadWebDocumentDataRequest String that contains web document URL and save options
	 * @param dataInterchangeFormat Can be either XML or JSON
	 * @throws java.security.InvalidKeyException If initialization fails because the provided key is null.
	 * @throws java.security.NoSuchAlgorithmException If the specified algorithm (HmacSHA1) is not available by any provider.
	 * @throws java.io.IOException If there is an IO error
	 * @return An object that contains URL to word document
	*/
	public static SaveResult convertWebPagesToWordDocument(String loadWebDocumentDataRequest, String dataInterchangeFormat) throws InvalidKeyException, NoSuchAlgorithmException, IOException {
		SaveResult saveResult = null;
		
		if(loadWebDocumentDataRequest == null || loadWebDocumentDataRequest.length() == 0) {
			throw new IllegalArgumentException("Load web document data request cannot be null or empty");
		}
		
		if(dataInterchangeFormat == null || dataInterchangeFormat.length() == 0) {
			throw new IllegalArgumentException("Data interchange format cannot be null or empty. Either it should be XML or JSON");
		}
		
		//build URL
		String strURL = WORD_URI + "loadWebDocument";
		//sign URL
		String signedURL = Utils.sign(strURL);
		InputStream responseStream = Utils.processCommand(signedURL, "POST", loadWebDocumentDataRequest, dataInterchangeFormat);
		String responseJSONString = Utils.streamToString(responseStream);
		
		//Parsing JSON
      	Gson gson = new Gson();
      	ConvertWebPageToWordDocumentResponse convertWebPageToWordDocResponse = gson.fromJson(responseJSONString, ConvertWebPageToWordDocumentResponse.class);
		if(convertWebPageToWordDocResponse.getCode().equals("200") && convertWebPageToWordDocResponse.getStatus().equals("OK")) {
			saveResult = convertWebPageToWordDocResponse.saveResult;
		}
		
		return saveResult;

	}
}

File: SDKs/Aspose_Cloud_SDK_For_Ruby/lib/Words/converter.rb
module Aspose
  module Cloud
    module Words
      class Converter
        def initialize(filename)
          @filename = filename
          raise 'filename not specified.' if filename.empty?
          @base_uri =  Aspose::Cloud::Common::Product.product_uri + '/words/' + @filename
        end

=begin
  Convert Word to Images, Multipage TIFF, HTML, PDF and other File Format using Cloud Storage
  @param string save_format Return file format.
=end
        def convert(save_format, folder_name = '', storage_type = 'Aspose', storage_name = '')
          raise 'save_format not specified.' if save_format.empty?

          str_uri = "#{@base_uri}?format=#{save_format}"
          str_uri = Aspose::Cloud::Common::Utils.append_storage(str_uri,folder_name,storage_name,storage_type)
          signed_str_uri = Aspose::Cloud::Common::Utils.sign(str_uri)
          response_stream = RestClient.get(signed_str_uri, {:accept=>'application/json'})

          valid_output = Aspose::Cloud::Common::Utils.validate_output(response_stream)

          if valid_output.empty?
            output_path = "#{Aspose::Cloud::Common::AsposeApp.output_location}#{Aspose::Cloud::Common::Utils.get_filename(@filename)}.#{save_format}"
            Aspose::Cloud::Common::Utils.save_file(response_stream,output_path)
          else
            valid_output
          end
        end

=begin
  Convert Word to other File Formats without using the Cloud Storage
  @param string input_file_path Path of the input file.
  @param string output_filename Name of the return file.
  @param string save_format Return file format.
=end
        def convert_local_file(input_file_path, output_filename, save_format)
          raise 'input_file_path not specified.' if input_file_path.empty?
          raise 'output_filename not specified.' if output_filename.empty?
          raise 'save_format not specified.' if save_format.empty?

          str_uri = "#{Aspose::Cloud::Common::Product.product_uri}/words/convert?format=#{save_format}"
          signed_str_uri = Aspose::Cloud::Common::Utils.sign(str_uri)
          response_stream = Aspose::Cloud::Common::Utils.upload_file_binary(input_file_path, signed_str_uri)

          valid_output = Aspose::Cloud::Common::Utils.validate_output(response_stream)

          if valid_output.empty?
            save_format = 'zip' if save_format.eql?('html')
            output_path = "#{Aspose::Cloud::Common::AsposeApp.output_location}#{Aspose::Cloud::Common::Utils.get_filename(output_filename)}.#{save_format}"
            Aspose::Cloud::Common::Utils.save_file(response_stream,output_path)
          end
          valid_output
        end

=begin
  Convert web pages to Word Documents
  @param xml str_xml Provide xml data.
=end
        def convert_web_pages(str_xml, folder_name = '', storage_type = 'Aspose', storage_name = '')
          raise 'str_xml not specified.' if str_xml.empty?
 
          str_uri = "#{Aspose::Cloud::Common::Product.product_uri}/words/loadWebDocument"
          str_uri = Aspose::Cloud::Common::Utils.append_storage(str_uri,folder_name,storage_name,storage_type)
          signed_str_uri = Aspose::Cloud::Common::Utils.sign(str_uri)

          response = Aspose::Cloud::Common::Utils.process_command(signed_str_uri,'POST','XML',str_xml)
          json = JSON.parse(response)
          
          if json['Code'] == 200
              filename = json['SaveResult']['DestDocument']['Href']
              Aspose::Cloud::Common::Utils.download_file(filename,filename,folder_name,storage_name,storage_type)
          end
        end
      end
    end
  end
end
